aiozk/recipes/base_watcher.py

```diff
@@ -28,21 +28,36 @@ def add_callback(self, path, callback):
         self.callbacks[path].add(callback)
 
         if path not in self.loops:
-            self.loops[path] = asyncio.ensure_future(self.watch_loop(path), loop=self.client.loop)
+            self.add_watch_loop(path)
 
     def remove_callback(self, path, callback):
         self.callbacks[path].discard(callback)
 
         if not self.callbacks[path]:
             self.callbacks.pop(path)
-            self.loops.pop(path).cancel()
+            self.discard_watch_loop(path)
+
+    def add_watch_loop(self, path):
+        log.debug('Add new watch loop for %s', path)
+        stop_event = asyncio.Event(loop=self.client.loop)
+        fut = asyncio.ensure_future(self.watch_loop(path, stop_event), loop=self.client.loop)
+        self.loops[path] = (fut, stop_event)
+
+    def discard_watch_loop(self, path):
+        log.debug('Discard watch loop of %s', path)
+        try:
+            fut, stop_event = self.loops.pop(path)
```

Comment:
We can ensure that path in loops and eliminate the whole try/except construction which will make code more clear

Reply:
Actually discard_watch_loop can be called more than once so we can not ensure that loops.pop(path) succeeds always.

Reply:
@cybergrind discard_watch_loop() method is called at 3 points: inside remove_callback(), under exception handler of watch_loop() and inside watch_loop().

Reply:
There is no async code there so it should work if rewritten to:
```python
    def discard_watch_loop(self, path):
        log.debug('Discard watch loop of %s', path)
        if path in self.loops:
            fut, stop_event = self.loops.pop(path)
            stop_event.set()
```
Just because it is impossible to get KeyError if you checked that it exists before accessing it.